hal/src/common/dmac/transfer.rs

```diff
@@ -0,0 +1,545 @@
+//! channel and the payload.
+//!
+//! # Trigger sources
+//!
+//! Most peripherals can issue triggers to a DMA channel. A software trigger is
+//! also available (see [`trigger`](Transfer::software_trigger)). See
+//! ATSAMD21 datasheet, table 19-8 for all available trigger sources.
+//!
+//! # Trigger actions
+//!
+//! Three trigger actions are available:
+//!
+//! * BLOCK: One trigger required for each block transfer. In the context of
+//!   this driver,
+//! one Transfer is equivalent to one Block transfer.
+//!
+//! * BEAT: One trigger required for each beat transfer. In the context of this
+//!   driver, the beat
+//! size will depend on the type of buffer used (8, 16 or 32 bits).
+//!
+//! * TRANSACTION: One trigger required for a full DMA transaction. this is
+//!   useful for circular
+//! transfers in the context of this driver. One trigger will set off the
+//! transaction, that will now run uninterrupted until it is stopped.
+
+use super::{
+    channel::{AnyChannel, Busy, Channel, ChannelId, Ready},
+    dma_controller::{ChId, DmaController, TriggerAction, TriggerSource},
+    BlockTransferControl, DmacDescriptor, DESCRIPTOR_SECTION,
+};
+use crate::typelevel::{Is, Sealed};
+use core::mem;
+use core::sync::atomic;
+use modular_bitfield::prelude::*;
+
+//==============================================================================
+// Beat
+//==============================================================================
+
+/// Useable beat sizes for DMA transfers
+#[derive(Clone, Copy, BitfieldSpecifier)]
+pub enum BeatSize {
+    /// Byte = [`u8`](core::u8)
+    Byte = 0x00,
+    /// Half word = [`u16`](core::u16)
+    HalfWord = 0x01,
+    /// Word = [`u32`](core::u32)
+    Word = 0x02,
+    _RESERVED = 0x03,
+}
+/// Convert 8, 16 and 32 bit types
+/// into [`BeatSize`](BeatSize)
+pub trait Beat: Sealed {
+    /// Convert to BeatSize enum
+    const BEATSIZE: BeatSize;
+}
+
+macro_rules! impl_beat {
+    ( $( ($Type:ty, $Size:ident) ),+ ) => {
+        $(
+            impl Sealed for $Type {}
+            impl Beat for $Type {
+                const BEATSIZE: BeatSize = BeatSize::$Size;
+            }
+        )+
+    };
+}
+
+impl_beat!(
+    (u8, Byte),
+    (i8, Byte),
+    (u16, HalfWord),
+    (i16, HalfWord),
+    (u32, Word),
+    (i32, Word),
+    (f32, Word)
+);
+
+//==============================================================================
+// Buffer
+//==============================================================================
+
+/// Buffer useable by the DMAC.
+pub unsafe trait Buffer {
+    type Beat: Beat;
+    /// Pointer to the buffer. If the buffer is incrementing, the address should
+    /// point to the last beat transfer in the block.
+    fn dma_ptr(&mut self) -> *mut Self::Beat;
+    /// Return whether the buffer pointer should be incrementing or not
+    fn incrementing(&self) -> bool;
+    /// Buffer length in beats
+    fn buffer_len(&self) -> usize;
```

Comment:
I had a realization about this a while back, but I forgot about it. I can't remember if I told you or not. If I did and you disregarded it, then please feel free to ignore it again.

I realized that inherent `impl` names never conflict with trait `impl` names. If there is an inherent `impl` name, it will always be used, even if there is a trait function with the same name. If you want to use the trait function in that case, you have to explicitly call it.

That works to your advantage here. I think you can use the name `len` here without worrying. If the type is a slice, then it has an inherent implementation of `len` which is already what you want. But if it's not a slice and doesn't already have a `len` function, you can add one through this trait. Personally, I like the symmetry of that approach, but there is maybe some possibility for confusion, so I can see arguments against it.

Reply:
I will keep `buffer_len` because this warning pops up when changing the trait method name to `len`:
``` rust
    |
201 |     fn len(&self) -> usize {
    |     ^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
202 |         self.len()
    |         ---------- recursive call site
    |
    = note: `#[warn(unconditional_recursion)]` on by default
    = help: a `loop` may express intention better if this is on purpose
```

Reply:
I was curious why this didn't work. I think I got it to work [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6200c0f4649fe6ca1484898a01812962).

That error must have come when implementing `Buffer` for arrays, right? I think I understand what's going on.

Arrays don't define their own `len` function. Instead, they use unsized coercion to access `len` defined for slices. When you call `self.len()` within the `Buffer` context, it ends up choosing the `Buffer` definition of `len` over the coerced definition from slices. The solution is to explicitly do the coercion with `<[T]>::len(self)`.